File: Tri.py
#! /usr/bin/python3

import numpy as np
# from mpl_toolkits.mplot3d import axes3d
# import matplotlib.pyplot as plt
import meshio


eps = 1.0E-6
inv4pi = 0.07957747155

class Tri:
    """ Class for handling triangular vortex elements """
    def __init__(self, p0, p1, p2, ncap, rc):
        self.p = np.zeros((3, 3))
        self.p[:, 0] = p0
        self.p[:, 1] = p1
        self.p[:, 2] = p2
        self.ncap = ncap
        self.rc = rc
        self.cp = (self.p[:, 0] + self.p[:, 1] + self.p[:, 2])/3.0

    def vind(self, p):
        vel = self.vindFilament(0, p) + self.vindFilament(1, p) + \
                self.vindFilament(2, p)
        return vel

    def vindFilament(self, filamentNum, p):
        if filamentNum == 0:
            p1 = self.p[:, 0]
            p2 = self.p[:, 1]
        elif filamentNum == 1:
            p1 = self.p[:, 1]
            p2 = self.p[:, 2]
        else:
            p1 = self.p[:, 2]
            p2 = self.p[:, 0]

        r1 = p-p1
        r2 = p-p2
        r0 = r1-r2

        r1xr2 = np.cross(r1, r2)
        r1xr2abs2 = np.dot(r1xr2, r1xr2)

        r1Unit = r1/np.linalg.norm(r1)
        r2Unit = r2/np.linalg.norm(r2)

        vel = 0.0
        if r1xr2abs2 > eps*eps:
            vel = (r1xr2*inv4pi*np.dot(r0, r1Unit-r2Unit))/ \
                    np.sqrt((self.rc*np.linalg.norm(r0))**4.0+r1xr2abs2**2.0)

        return vel

class Surface:
    def __init__(self, rc, STLfilename, isClosed):
        stlData = meshio.read(STLfilename)
        self.mesh = stlData
        self.nElem = stlData.cells[0].data.shape[0]
        self.gamma = np.zeros(self.nElem)
        self.ele = []
        self.isClosed = isClosed
        for i in range(self.nElem):
            vtx = stlData.cells_dict['triangle'][i, :]
            p0 = stlData.points[vtx[0], :]
            p1 = stlData.points[vtx[1], :]
            p2 = stlData.points[vtx[2], :]

            try:
                ncap = stlData.cell_data['facet_normals'][0][i, :]
            except KeyError:
                ncap = np.cross(p1-p0, p2-p1)

            ncap = ncap/np.linalg.norm(ncap)
            self.ele.append(Tri(p0, p1, p2, ncap, rc))
        print('Computing AIC ...')
        self.aic = self.getAIC()

    def getAIC(self):
        aic = np.zeros((self.nElem, self.nElem))
        for i in range(self.nElem):
            for j in range(self.nElem):
                aic[i, j] = np.dot(self.ele[j].vind(self.ele[i].cp), \
                                   self.ele[i].ncap)
        if self.isClosed:
            # Assumes first element has gamma = 0.0
            return aic[1:, 1:]
        else:
            return aic

    def vind(self, p):
        vel = 0.0
        for i in range(self.nElem):
            vel += self.ele[i].vind(p)*self.gamma[i]
        return vel

    def getRHS(self, vinf):
        RHS = np.zeros((self.nElem))

        for i in range(self.nElem):
            RHS[i] = np.dot(np.array(vinf),  self.ele[i].ncap)

        if self.isClosed:
            return RHS[1:]
        else:
            return RHS

    def assignGamma(self, gamma):
        if len(gamma) == self.nElem-1:
            self.gamma[0] = 0.0
            for i in range(1, self.nElem):
                self.gamma[i] = gamma[i-1]
        else:
            self.gamma = gamma
        self.mesh.cell_data['gamma'] = self.gamma

    def writeVTK(self, gamma, VTKfilename):
        self.mesh.write(VTKfilename)

    def writeGrid(self, vinf, xlim, ylim, zlim, nvec, TECfilename):
        x = np.linspace(xlim[0], xlim[1], nvec[0])
        y = np.linspace(ylim[0], ylim[1], nvec[1])
        z = np.linspace(zlim[0], zlim[1], nvec[2])

        xg, yg, zg = np.meshgrid(x, y, z)
        u = np.zeros((nvec[0], nvec[1], nvec[2]))
        v = np.zeros((nvec[0], nvec[1], nvec[2]))
        w = np.zeros((nvec[0], nvec[1], nvec[2]))

        for k in range(nvec[2]):
            for j in range(nvec[1]):
                for i in range(nvec[0]):
                    u[i,j,k], v[i,j,k], w[i,j,k] = \
                            self.vind([xg[i, j, k], \
                                       yg[i, j, k], \
                                       zg[i, j, k]])

        u += vinf[0]
        v += vinf[1]
        w += vinf[2]

        # Write to tec file
        with open(TECfilename, "w") as fh:
            fh.write('TITLE = "Grid"\n')
            fh.write('VARIABLES = "X" "Y" "Z" "U" "V" "W"\n')
            fh.write('ZONE I='+ str(nvec[0]) + ' J=' + str(nvec[1])+ \
                    ' K=' + str(nvec[2]) + ' DATAPACKING=BLOCK')

            for dt in [xg, yg, zg, u, v, w]:
                fh.write("\n")
                for k in range(nvec[2]):
                    for j in range(nvec[1]):
                        for i in range(nvec[0]):
                            fh.write("%e\n" % dt[i, j, k])


body = Surface(1E-6, "sphere.stl", isClosed = True)
vinf = [1,0,0]
RHS = body.getRHS(vinf)
print('Computing solution ...')
gamma = np.linalg.solve(body.aic, RHS)
body.assignGamma(gamma)
body.writeVTK(gamma, "body.vtk")
# body.writeGrid(vinf, [0,5], [0,5], [0,5], [30, 30, 30], 'grid.tec')
body.writeGrid(vinf, [-250, 260], [-560,-70], [-85,420], [30, 30, 30], 'grid.tec')

File: Tri.jl
using VSPGeom
using VSPGeom.WriteVTK
using LinearAlgebra

inv4pi = 0.07957747155
eps2 = eps()^2

""" Class for handling triangular vortex elements """
struct Tri{TF}
    p::Matrix{TF}
    ncap::Vector{TF}
    rc::TF
    cp::Vector{TF}
end

""" Constructor """
function Tri(p, rc)
    ncap = cross(p[:, 2]-p[:, 1], p[:, 3]-p[:, 2])
    ncap .= ncap/norm(ncap)

    cp = zeros(3)
    cp .= sum(p; dims=2) ./ 3
    return Tri(p, ncap, rc, cp)
end

function Tri(p1::Vector{TF}, p2::Vector{TF}, p3::Vector{TF}, rc::TF) where (TF<:Float64)
    p = zeros(3, 3)
    p[:, 1] = p1
    p[:, 2] = p2
    p[:, 3] = p3

    return Tri(p, rc)
end

""" Induced velocity """
function vind(self::Tri, p)
    return vindFil(self, 1, p) + vindFil(self, 2, p) + vindFil(self, 3, p)
end


""" Induced velocity by filament """
function vindFil(self::Tri, ifil::Int, p::Vector{Float64})
    if ifil < 3
        p1 = self.p[:, ifil]
        p2 = self.p[:, ifil+1]
    else
        p1 = self.p[:, 3]
        p2 = self.p[:, 1]
    end

    r1 = p-p1
    r2 = p-p2
    r0 = r1-r2

    r1xr2 = cross(r1, r2)
    r1xr2abs2 = dot(r1xr2, r1xr2)

    r1Unit = r1/norm(r1)
    r2Unit = r2/norm(r2)

    vel = 0.0
    if r1xr2abs2 > eps2
        vel = (r1xr2*inv4pi*dot(r0, r1Unit-r2Unit)) / sqrt((self.rc*norm(r0))^4.0+r1xr2abs2^2.0)
    end
    return vel
end

struct Surface{TI, TF}
    mesh::TriMesh
    ele::Vector{Tri}
    nElem::TI
    gamma::Vector{TF}
    isClosed::Bool
    aic::Matrix{TF}
end

""" Constructor """
function Surface(filename::String; rc=1e-6, isClosed=true, tol=1e-6)
    geom = readSTL(filename)
    mesh = geom[1]
    nElem = length(mesh.cells)
    gamma = zeros(nElem)

    ele = Vector{Tri}(undef, nElem)
    for i in 1:nElem
        ele[i] = Tri(mesh.points[mesh.cells[i]][1],
                     mesh.points[mesh.cells[i]][2],
                     mesh.points[mesh.cells[i]][3], rc)
    end

    # Compute aic matrix
    aic = zeros(nElem, nElem)

    for j in 1:nElem
        for i in 1:nElem
            aic[i, j] = dot(vind(ele[j], ele[i].cp), ele[i].ncap)
        end
    end

    return Surface(mesh, ele, nElem, gamma, isClosed, aic)
end

function vind(self::Surface, p)
    vel = 0.0
    for i in range(1:self.nElem)
        vel += vind(self.ele[i], p) .* self.gamma[i]
    end
    return vel
end

function getRHS(self::Surface, vinf)
    RHS = zeros(self.nElem)

    for i in 1:self.nElem
        RHS[i] = -1 .* dot(vinf, self.ele[i].ncap)
    end

    return RHS
end

function writeMesh(s::Surface, filename::String; vinf=zeros(3))
    points, cells = getVTKElements(s.mesh)
    ncap = zeros(3, s.nElem)
    cp = zeros(3, s.nElem)
    id = 1:s.nElem
    for i = 1:s.nElem
        ncap[:, i] = s.ele[i].ncap
        cp[:, i] = s.ele[i].cp
    end
    vtk_grid(filename, points, cells) do vtk
        vtk["gamma"] = s.gamma
        vtk["cell_id"] = id
        vtk["cp", VTKCellData()] = cp
        vtk["ncap", VTKCellData()] = ncap
        vtk["vel", VTKCellData()] = repeat(vinf, outer=(1, s.nElem))
    end
end

function solve(aic, RHS; isClosed=true)
    if isClosed
        return aic[2:end, 2:end] \ RHS[2:end]
    else
        return aic \ RHS
    end
end

function assignGamma!(s::Surface, gamma)
    idxStart = 1
    if length(gamma) == s.nElem-1
        s.gamma[1] = 0.0
        idxStart = 2
    end
    s.gamma[idxStart:end] .= gamma
end

# Main program
body = Surface("octa.stl")
vinf = [1,0,0]
RHS = getRHS(body, vinf)
println("Computing solution ...")
@show gamma = solve(body.aic, RHS; isClosed=body.isClosed)
assignGamma(body, gamma)
writeMesh(body, "out"; vinf=vinf)
# # body.writeGrid(vinf, [0,5], [0,5], [0,5], [30, 30, 30], 'grid.tec')
# body.writeGrid(vinf, [-250, 260], [-560,-70], [-85,420], [30, 30, 30], 'grid.tec')
